File: TCDNew-master/src/ui.android/lib/services/PurchaseTaskStateView.dart
import 'dart:async';

import 'package:tiq/models/dtos/PurchaseTaskLineDto.dart';
import 'package:tiq/models/dtos/PurchaseTaskPalletDto.dart';
import 'package:tiq/models/dtos/PurchaseTaskUpdateDto.dart';


class PurchaseTaskStateView {
  String purchaseTaskId;
  int purchaseTaskVersion = 0;
  List<PurchaseTaskLineDto> _purchaseTaskLines = List<PurchaseTaskLineDto>();
  List<PurchaseTaskPalletDto> _purchaseTaskPallets = List<PurchaseTaskPalletDto>();
  String _currentPalletCode = '';

  PurchaseTaskStateView(this.purchaseTaskId);

  int getDifferencesCount() {
    final count = _purchaseTaskLines.where((e) => e.quantity != e.state.qtyFull).length;

    return count;
  }

  int getTotalCount() {
    return _purchaseTaskLines.length;
  }

  int getAcceptedCount() {
    final count = _purchaseTaskLines.where((e) => e.quantity == e.state.qtyFull).length;

    return count;
  }

  void setCurrentPalletCode(String palletCode) {
    this._currentPalletCode = palletCode;
  }

  String getCurrentPalletCode() {
    return this._currentPalletCode ?? '';
  }

  String getCurrentPalletABC() {
    final purchaseTaskPallet = _purchaseTaskPallets.firstWhere(
        (e) => e.code == _currentPalletCode,
        orElse: () => null);
    return purchaseTaskPallet != null ? purchaseTaskPallet.abc : "";
  }

  bool isPalletABCUsed(String abc) {
    final purchaseTaskPallet = _purchaseTaskPallets.firstWhere(
        (e) => e.abc.toLowerCase().contains(abc.trim().toLowerCase()),
        orElse: () => null);
    return purchaseTaskPallet != null;
  }

  bool isPalletABCEmpty(String abc) {
    return abc.trim() == "";
  }

  bool canPalletAcceptProductAbc(String palletABC, String goodABC) {
    return ((palletABC.toLowerCase() == "") || (palletABC.toLowerCase().contains(goodABC.trim().toLowerCase())));
  }

  void purchaseTaskUpdateApply(PurchaseTaskUpdateDto purchaseTaskUpdate) {
    if (purchaseTaskUpdate.purchaseTaskId != this.purchaseTaskId) return;
    // if (purchaseTaskUpdate.purchaseTaskVersion <= this.purchaseTaskVersion) return;

    for(final lineUpdated in purchaseTaskUpdate.linesUpdated) {
      final lineLocal = this._purchaseTaskLines.firstWhere(
          (e) => e.id == lineUpdated.id,
          orElse: () => null);
      if (lineLocal != null) {
        lineLocal.quantity = lineUpdated.quantity;
        lineLocal.state.expirationDate = lineUpdated.state.expirationDate;
        lineLocal.state.expirationDaysPlus = lineUpdated.state.expirationDaysPlus;
        lineLocal.state.qtyNormal = lineUpdated.state.qtyNormal;
        lineLocal.state.qtyBroken = lineUpdated.state.qtyBroken;
        lineLocal.product.id = lineUpdated.product.id;
        lineLocal.product.name = lineUpdated.product.name;
        lineLocal.product.abc = lineUpdated.product.abc;
        lineLocal.product.barcodes = lineUpdated.product.barcodes;
      } else {
        this._purchaseTaskLines.add(lineUpdated);
      }
    }
    for(final palleteUpdated in purchaseTaskUpdate.palletsUpdated) {
      final palleteLocal = this._purchaseTaskPallets.firstWhere(
              (e) => e.code == palleteUpdated.code,
              orElse: () => null);
      if (palleteLocal != null) {
        palleteLocal.abc = palleteUpdated.abc;
      } else {
        this._purchaseTaskPallets.add(palleteUpdated);
      }
    }

    this.purchaseTaskVersion = purchaseTaskUpdate.purchaseTaskVersion;
  }

  Future<List<PurchaseTaskLineDto>> getLinesDataSource() async {
    return this._purchaseTaskLines;
  }
}

File: TCDNew-master/src/ui.wince/Cen.Wms.Client/Services/PurchaseTaskStateView.cs
using System.Collections.Generic;
using System.ComponentModel;
using System.Linq;
using System.Windows.Forms;
using Cen.Wms.Client.Models.Dtos;

namespace Cen.Wms.Client.Services
{
    public class PurchaseTaskStateView
    {
        private readonly BindingList<PurchaseTaskLineDto> _purchaseTaskLines;
        private readonly List<PurchaseTaskPalletDto> _purchaseTaskPallets;
        private string _currentPalletCode = string.Empty;

        public readonly string PurchaseTaskId;
        public long PurchaseTaskVersion = 0;

        public PurchaseTaskStateView(string purchaseTaskId)
        {
            this.PurchaseTaskId = purchaseTaskId;
            _purchaseTaskLines = new BindingList<PurchaseTaskLineDto>();
            _purchaseTaskPallets = new List<PurchaseTaskPalletDto>();
        }

        public int GetDifferencesCount()
        {
            return _purchaseTaskLines.Count(line => (line.Quantity != line.State.QtyFull));
        }

        public int GetTotalCount()
        {
            return _purchaseTaskLines.Count();
        }

        public int GetAcceptedCount()
        {
            return _purchaseTaskLines.Count(line => (line.Quantity == line.State.QtyFull));
        }

        public void SetCurrentPalletCode(string palletCode)
        {
            _currentPalletCode = palletCode;
        }

        public string GetCurrentPalletCode()
        {
            return _currentPalletCode;
        }

        public string GetCurrentPalletABC()
        {
            var purchaseTaskPallet = _purchaseTaskPallets.FirstOrDefault(e => e.Code == _currentPalletCode);
            return purchaseTaskPallet != null ? purchaseTaskPallet.ABC : string.Empty;
        }

        public bool IsPalletABCUsed(string abc)
        {
            var purchaseTaskPallet = _purchaseTaskPallets.FirstOrDefault(e => e.ABC.Contains(abc.Trim().ToLower()));
            return purchaseTaskPallet != null;
        }

        public bool IsPalletABCEmpty(string abc)
        {
            return abc.Trim() == string.Empty;
        }

        public bool CanPalletAcceptProductABC(string palletABC, string goodABC)
        {
            return ((palletABC.ToLower() == "") || (palletABC.ToLower().Contains(goodABC.Trim().ToLower())));
        }

        public void PurchaseTaskUpdateApply(PurchaseTaskUpdateDto purchaseTaskUpdate)
        {
            if (purchaseTaskUpdate.PurchaseTaskId != this.PurchaseTaskId) return;
            // if (purchaseTaskUpdate.PurchaseTaskVersion <= this.PurchaseTaskVersion) return;

            foreach(var lineUpdated in purchaseTaskUpdate.LinesUpdated)
            {
                var purchaseTaskLineIndex = -1;
                for (var i = 0; i < this._purchaseTaskLines.Count; i++)
                {
                    if (_purchaseTaskLines[i].Id == lineUpdated.Id)
                    {
                        purchaseTaskLineIndex = i;
                        break;   
                    }
                }
                var lineLocal = purchaseTaskLineIndex != -1 ? this._purchaseTaskLines[purchaseTaskLineIndex] : null;
                if (lineLocal != null)
                {
                    lineLocal.Quantity = lineUpdated.Quantity;
                    lineLocal.State.ExpirationDate = lineUpdated.State.ExpirationDate;
                    lineLocal.State.ExpirationDaysPlus = lineUpdated.State.ExpirationDaysPlus;
                    lineLocal.State.QtyNormal = lineUpdated.State.QtyNormal;
                    lineLocal.State.QtyBroken = lineUpdated.State.QtyBroken;
                    lineLocal.Product.Id = lineUpdated.Product.Id;
                    lineLocal.Product.Name = lineUpdated.Product.Name;
                    lineLocal.Product.Abc = lineUpdated.Product.Abc;
                    lineLocal.Product.Barcodes = lineUpdated.Product.Barcodes;
                    this._purchaseTaskLines.ResetItem(purchaseTaskLineIndex);
                }
                else
                {
                    this._purchaseTaskLines.Add(lineUpdated);
                }
            }
            foreach (var palletUpdated in purchaseTaskUpdate.PalletsUpdated)
            {
                var palletLocal = this._purchaseTaskPallets.FirstOrDefault(e => e.Code == palletUpdated.Code);
                if (palletLocal != null)
                {
                    palletLocal.ABC = palletUpdated.ABC;
                }
                else
                {
                    this._purchaseTaskPallets.Add(palletUpdated);
                }
            }

            this.PurchaseTaskVersion = purchaseTaskUpdate.PurchaseTaskVersion;
        }

        public BindingSource GetLinesDataSource()
        {
            _purchaseTaskLines.RaiseListChangedEvents = true;
            return new BindingSource { DataSource = _purchaseTaskLines };
        }
    }
}
